src/hera/workflows/io/_io_mixins.py

```diff
@@ -39,21 +40,29 @@
     BaseModel = object  # type: ignore
 
 
+@contextmanager
+def no_active_context() -> Iterator[None]:
+    pieces = _context.pieces
+    _context.pieces = []
+    try:
+        yield
+    finally:
+        _context.pieces = pieces
+
+
 class InputMixin(BaseModel):
     def __new__(cls, **kwargs):
-        if _context.declaring:
+        if _context.active:
             # Intercept the declaration to avoid validation on the templated strings
-            # We must then turn off declaring mode to be able to "construct" an instance
+            # We must then disable the active context to be able to "construct" an instance
             # of the InputMixin subclass.
-            _context.declaring = False
-            instance = cls.construct(**kwargs)
-            _context.declaring = True
-            return instance
+            with no_active_context():
+                return cls.construct(**kwargs)
```

Comment:
I don't understand hera's context management 100%, so deferring mixins' review to others.
For me, it seems refine the context state checking, and looks fine.